[AArch64] Add updated FEAT_SVE_B16B16 and begin replacement of 'b16b16' flag

SpencerAbson opened the pull request:
This patch adds FeatureSVEB16B16 to the AArch64 backend in order to represent the new behavior of FEAT_SVE_B16B16 (as described in the latest [Armv9.4 extensions documentation](https://developer.arm.com/documentation/109697/0100/Feature-descriptions/The-Armv9-4-architecture-extension?lang=en#md461-the-armv94-architecture-extension__FEAT_SVE_B16B16)) as well as a 'sve-b16b16' flag to enable it. 

The predication of non-widening SVE BFloat16 instructions has changed to require this feature, instead of the previously required and soon-to-be-removed FeatureB16B16 which is enabled by the 'b16b16' flag. Therefore, this change weakens the 'b16b16' flag in favour of 'sve-b16b16'. Existing tests that are effected by this have been modified to use and/or expect 'sve-b16b16', and new tests have been added to verify the behavior and implementation of 'sve-b16b16'. 

This patch is in response to the response to the following changes.

The architecture features previously enabled by FEAT_SVE_B16B16 have been relaxed such that it now implements:
      - With FEAT_SVE2 : SVE non-widening BFloat16 instructions in Non-streaming SVE mode
      - With FEAT_SME2: SVE non-widening BFloat16 instructions when the PE is in Streaming SVE mode and SME
        Z-targeting multi-vector non-widening BFloat16 instructions.
      - **It no longer implements** SME ZA-targeting non-widening BFloat16 instructions.      

The SME ZA-targeting non-widening BFloat16 instructions are implemented by the new FEAT_SME_B16B16, **this patch does not change how this architecture feature is enabled** ('+b16b16+sme2'). Only those that are implemented by FEAT_SVE_B16B16 have been changed to require 'sve-b16b16' instead of 'b16b16'.

New flags must be created to represent FEAT_SVE_B16B16 and FEAT_SME_B16B16:
      - 'sve-b16b16' enables the updated FEAT_SVE_B16B16 (described here)
      - 'sme-b16b16' will enable the new FEAT_SME_B16B16
      - **This patch includes 'sve-b16b16' only**
   
A future patch will add 'sme-b16b16', SME ZA-targeting non-widening BFloat16 instructions would then be guarded by '+sme-b16b16+sme2', and 'b16b16' can be removed.

SpencerAbson committed:
Add FEAT_SVE_B16B16 and update tests under B16B16

Implement the new FEAT_SVE_B16B16 semantics and flag ('sve-b16b16') to
AArch64; add llvm/MC directive and clang/Sema tests to verify the use of
this flag; modiy the llvm/MC, Clang/CodeGen, Clang/Sema, and
llvm/CodeGen tests that are affected by the weakening of the soon-to-be
-deprecated 'b16b16' flag.

- Changes to Clang frontend
 - clang/include/clang/Basic/arm_sve.td
	- SVE single-vector intrinsics are guarded by +sve-b16b16+sve2 or
          +sve-b16b16+sme2 (non-streaming mode and streaming mode)
        - Z-targeting multi-vector instrinsics are gated on +sve-b16b16+sme2
          (svclamp_single_[x2/x4]) and are streaming mode only

- Changes to LLVM AArch64 backend
 - llvm/lib/Target/AArch64/AArch64Features.td
	- Create FeatureSVEB16B16
        - Change FeatureB16B16 to separate it from the new meaning of
	  FEAT_SVE_B16B16, it is now weakened to the role of the incoming
	  FEAT_SME_B16B16
 - llvm/lib/Target/AArch64/AArch64InstrInfo.td
	- Create the HasSVEB16B16 predicate, requires that the subtarget
	 has sve-b16b16 enabled
 - llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td
	- Predicate Z-targeting multi-vector BFloat to BFloat instructions
	  on HasSME2 and HasSVEB1616 (replace HasB16B16)
	- ZA-targeting instructions are separated and remain predicated
	  under HasB16B16 before it is replaced by incoming HasSMEB16B16
 - llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td
	- Replace HasB16B16 with HasSVEB16B16 in predication of SVE2 single
	  vector BFloat to BFloat instructions
 - llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp
	- Add the new sve-b16b16 flag mapping to FeatureSVEB16B16

- Changes to LLVM unit tests
 - llvm/unittests/TargetParser/TargetParserTest.cpp
	- Add new sve-b16b16 flag to existing target parser tests

- Added tests
 - clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_b16b16_streaming.cpp
	- To ensure that +sve-b16b16+sme2 enables SVE2 single vector BFloat
	to BFloat intrinsics in streaming mode
 - clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_b16b16.cpp
	- To ensure that +sve-b16b16+sve2 enables SVE2 single vector BFloat
	to BFloat intrinsics in non-streaming mode
 - llvm/test/MC/AArch64/SVE2p1
	- To ensure that SVE2 single vector BFloat to BFloat instructions are
	enabled by +sve-b16b16+sve2, and that this feature is removed by the
	prescence of +nosve-b16b16
 - llvm/test/MC/AArch64/SME2p1
	- To ensure that Z-targeting multi vector BFloat to BFloat instructions are
	enabled by +sve-b16b16+sme2, and that this feature is removed by the
	prescence of +nosve-b16b16

- Modified tests
	- All CodeGen, Semantic, and MC tests that are effected by the weakening
	 of +b16b16 have been modified to supply and/or expect +sve-b16b16.

SpencerAbson committed:
Move FeatureSVEB16B16 to Armv9.4 features

CarolineConcatto commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -2172,7 +2172,7 @@ let SVETargetGuard = InvalidMode, SMETargetGuard = "sme2" in {
 }
 
-let SVETargetGuard = InvalidMode, SMETargetGuard = "sme2,b16b16"in {
+let SVETargetGuard = InvalidMode, SMETargetGuard = "sme2,sve-b16b16"in {
```
It is missing the multi vector targeting Z. For some reason they are not under b16b16.
But they were added in this commit f7392f40f3f6d5d4f.
I can see them being changes in LLVM in this patch
// SME2 Z-targeting non-widening BFloat16 instructions
let Predicates = [HasSME2, HasSVEB16B16] in {

SpencerAbson replied:
Thanks for pointing this out.

CarolineConcatto reviewed (commented):
Thank you Spencer for the patch.
I believe we missed the  multi vector instructions targeting BF16.

SpencerAbson commented on clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_b16b16.cpp:
```diff
@@ -0,0 +1,49 @@
+// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve -target-feature +sve2 -verify -verify-ignore-unexpected=error,note -emit-llvm -o - %s
```
Will do :+1:

SpencerAbson committed:
Add SME2 Z-t b16-to-b16 intrinsics under sve-b161b6

- Predicate the SME2 Z-targeting b16-to-b16 intrinsics under FEAT_SME2,
  FEAT_SVE_B16B16.
- Add tests in ./clang/test/Sema/aarch64-sme2-intrinsics/acle_sme2_b16b16.cpp
  to verify this change.

- Combine streaming/non-streaming SVE mode tests into a single file,
  ./clang/test/Sema/aarch4-sve2p1-intrinsics/acle_sve2p1_b16b16.
- Delete ./clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_b16b16_streaming.cpp
  (redundant).

CarolineConcatto commented on clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfclamp.c:
```diff
@@ -1,10 +1,10 @@
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
```
Not related to you patch, but the instructions that you changed  now only require sve2.
I believe you can change the run line to be only sve2.
The only ones that are sve2.1 are in section
SVE2.1 instruction intrinsics in the ACLE https://github.com/ARM-software/acle/blob/main/main/acle.md

CarolineConcatto commented on clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmla_lane.c:
```diff
@@ -1,10 +1,10 @@
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
```
Same here.
s/sve2p1/sve2/
Please!

CarolineConcatto commented on clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmls_lane.c:
```diff
@@ -1,10 +1,10 @@
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
```
Same here.
s/sve2p1/sve2/
Please!

CarolineConcatto commented on clang/test/Sema/aarch64-sve2-intrinsics/acle_sve2_bfloat.cpp:
```diff
@@ -26,4 +26,4 @@ void test_bfloat(const bfloat16_t *const_bf16_ptr, svbfloat16_t bf16, svbfloat16
   // overload-error@+1 {{'svwhilewr' needs target feature (sve2,bf16)|(sme,bf16)}}
   SVE_ACLE_FUNC(svwhilewr,_bf16,,)(const_bf16_ptr, const_bf16_ptr);
-}
+}
```
Unrelated change

CarolineConcatto reviewed (approved):
Thank you Spencer for the work.
I left some comments about the sve2.1 feature flags being used for the sve2. But I dont think you need to change in this patch. Because it is unrelated
Maybe you create another patch a NFC to move all the sve2p1 tests and run lines to be only SVE2?
I believe when we changed the feature flag we forgot to move the tests and also to fix all run lines.

SpencerAbson committed:
Remove unreleated test change

vfdff commented:
hi, SpencerAbson
    How to enable the behavior of **FEAT_SVE_B16B16** after this commit  ? I try **clang -march=armv8.5a+sve2+sve-b16b16**, but it seems doesn't take effect. https://gcc.godbolt.org/z/sadrshqvM 
  > Or the only method to effect is **-march=armv8.5a+sve2+bf16 -O3 -Xclang -target-feature -Xclang +sve-b16b16** ?

SpencerAbson commented:
Hi @vfdff

You were correct the first time, `-march=armv8.5a+sve2+sve-b16b16` will enable the instructions (and intrinsics) under [FEAT_SVE_B16B16](https://developer.arm.com/documentation/109697/0100/Feature-descriptions/The-Armv9-4-architecture-extension?lang=en#md461-the-armv94-architecture-extension__FEAT_SVE_B16B16).

However, please refer to the [ACLE ](https://github.com/ARM-software/acle/blob/latest-release/main/acle.md) concerning the use of the `__bf16` type. 
> The __bf16 type is only available when the __ARM_BF16_FORMAT_ALTERNATIVE feature macro is defined. When it is available it can only be used by the ACLE intrinsics ; it cannot be used with standard C operators. It is expected that arithmetic using standard C operators be used using a single-precision floating point format and the value be converted to __bf16 when required using ACLE intrinsics. 

Please bare in mind that the B16B16 support is in alpha state, and may be changed in the future.  If you would like more information about Arm intrinsics support for Bfloat16, feel free to reach out to me personally (spencer.abson@arm.com). It might also be useful to have a look at recent changes to the [main branch of the ACLE](https://github.com/ARM-software/acle/blob/main/main/acle.md) to get an idea about how the interface should look in the near future.

Thanks
